src/apps/ipv4/arp.lua

```diff
@@ -0,0 +1,271 @@
+                          src_mac, src_ipv4, dst_mac, dst_ipv4)
+end
+
+local function is_arp(p)
+   if p.length < ether_arp_header_len then return false end
+   local h = ffi.cast(ether_arp_header_ptr_t, p.data)
+   return ntohs(h.ether.type) == ether_type_arp
+end
+
+local function ipv4_eq(a, b) return C.memcmp(a, b, 4) == 0 end
+local function mac_eq(a, b)  return C.memcmp(a, b, 6) == 0 end
+
+local function copy_mac(src)
+   local dst = ffi.new('uint8_t[6]')
+   ffi.copy(dst, src, 6)
+   return dst
+end
+
+local function random_locally_administered_unicast_mac_address()
+   local mac = lib.random_bytes(6)
+   -- Bit 0 is 0, indicating unicast.  Bit 1 is 1, indicating locally
+   -- administered.
+   mac[0] = bit.lshift(mac[0], 2) + 2
+   return mac
+end
+
+ARP = {}
+local arp_config_params = {
+   -- Source MAC address will default to a random address.
+   self_mac = { default=false },
+   -- Source IP is required though.
+   self_ip  = { required=true },
+   -- The next-hop MAC address can be statically configured.
+   next_mac = { default=false },
+   -- But if the next-hop MAC isn't configured, ARP will figure it out.
+   next_ip  = { default=false }
+}
+
+function ARP:new(conf)
+   local o = lib.parse(conf, arp_config_params)
+   if not o.self_mac then
+      o.self_mac = random_locally_administered_unicast_mac_address()
+   end
+   if not o.next_mac then
+      assert(o.next_ip, 'ARP needs next-hop IPv4 address to learn next-hop MAC')
+      o.arp_request_pkt = make_arp_request(o.self_mac, o.self_ip, o.next_ip)
+      self.arp_request_interval = 3 -- Send a new arp_request every three seconds.
+   end
+   return setmetatable(o, {__index=ARP})
+end
+
+function ARP:maybe_send_arp_request (output)
+   if self.next_mac then return end
+   self.next_arp_request_time = self.next_arp_request_time or engine.now()
+   if self.next_arp_request_time <= engine.now() then
+      print(("ARP: Resolving '%s'"):format(ipv4:ntop(self.next_ip)))
+      self:send_arp_request(output)
+      self.next_arp_request_time = engine.now() + self.arp_request_interval
+   end
+end
+
+function ARP:send_arp_request (output)
+   transmit(output, packet.clone(self.arp_request_pkt))
+end
+
+function ARP:push()
+   local isouth, osouth = self.input.south, self.output.south
+   local inorth, onorth = self.input.north, self.output.north
+
+   self:maybe_send_arp_request(osouth)
+
+   for _ = 1, link.nreadable(isouth) do
+      local p = receive(isouth)
+      if p.length < ether_header_len then
+         -- Packet too short.
+         packet.free(p)
+      elseif is_arp(p) then
+         local h = ffi.cast(ether_arp_header_ptr_t, p.data)
+         if (ntohs(h.arp.htype) ~= arp_htype_ethernet or
+             ntohs(h.arp.ptype) ~= arp_ptype_ipv4 or
+             h.arp.hlen ~= 6 or h.arp.plen ~= 4) then
+            -- Ignore invalid packet.
+         elseif ntohs(h.arp.oper) == arp_oper_request then
+            if self.self_ip and ipv4_eq(h.arp.tpa, self.self_ip, 4) then
```

Comment:
`ipv4_eq` always compares 4 bytes, so the third parameter in the function call can be removed.